### PR TITLE
[TT-3674] Added support of config_data in API Definition

### PR DESCRIPTION
**Pull request by komalsukhani**
## Description

- Exposed `config_data` field of Apidefinition.
- Added sample yaml at `config/samples/config_data_virtual_endpoint.yaml`. 
- Updated Apidefinition doc.

## Related Issue

#356 

## Test Coverage For This Change

Tested locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [x] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [x] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [x] I have updated the documentation accordingly.
- [x] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor` in applicable directories.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `gofmt -s -w .`
  - [x] `go vet ./...`
  - [x] `golangci-lint run`

### CODE REVIEW

**Review by jlucktay — COMMENTED** [review]
> Added sample yaml at `config/samples/config_data_virtual_endpoint.yaml`.

I can't see this sample YAML file in the PR..?

**Comment by komalsukhani** [comment]
> > Added sample yaml at `config/samples/config_data_virtual_endpoint.yaml`.
> 
> I can't see this sample YAML file in the PR..?

Sorry. Just pushed in.

File: docs/api_definitions.md [hunk 1]
```diff
@@ -65,7 +65,7 @@ An API Definition describes the configuration of an API. It instructs Tyk Gatewa
 | Feature | Supported | Comments |
 | ----------- | --------- | --------- |
 | API Tagging | ✅ | - |
-| Config Data | ❌ | Not Implemented |
+| Config Data | ✅ | - |
```
Comment (asoorm) [inline docs/api_definitions.md:68]: Provide link to the sample `config_data_virtual_endpoint.yaml`

File: api/model/api.go [hunk 2]
```diff
@@ -708,7 +714,9 @@ type APIDefinitionSpec struct {
 	// Context Variables are available in the url rewriter, modify headers and body transforms.
 	EnableContextVars bool `json:"enable_context_vars,omitempty"`
 
-	//ConfigData MapStringInterface `json:"config_data"`
+	// +optional
+	// +nullable
+	ConfigData *ConfigDataType `json:"config_data"`
```
Comment (asoorm) [inline api/model/api.go:719]: What is ConfigData? what does it do?

**Comment by asoorm** [comment]
We can use this same mechanism to sort https://github.com/TykTechnologies/tyk-operator/issues/59

File: api/model/api.go [hunk 3]
```diff
@@ -1049,3 +1058,30 @@ func (ls ListAPIOptions) Params() url.Values {
...
+
+	in.Object = m
+
+	return nil
+}
+
+func (in *ConfigDataType) MarshalJSON() (data []byte, err error) {
+	m := in.Object
+
+	return json.Marshal(m)
```
Comment (buraksekili) [inline api/model/api.go:1086]: Does `in.Object` need to be assigned to a new `m` variable? If not, we might return as:
```go
return json.Marshal(in.Object)
```

**Review by buraksekili — APPROVED** [review]
Thanks @komalsukhani

File: api/model/api.go (hunk 3 above)
Comment (komalsukhani) [reply api/model/api.go]: Nice spot. Fixed in `5f25444`

**Pull request merged**